server/eclipse-project/src/main/java/us/freeandfair/corla/controller/ContestCounter.java

```diff
@@ -51,23 +51,69 @@ private ContestCounter() {
       .collect(Collectors.toList());
   }
 
+  /**
+   * Calculates all the pairwise margins - like a cross product - using
+   * the vote totals. When there are no losers, all margins are zero.
+   *
+   * @param winners those who won the contest
+   * @param losers those who did not win the contest
+   * @param voteTotals a map of choice name to number of votes received
+   * @return a Set of Integers representing all margins between winners
+   * and losers
+   */
+  public static Set<Integer> pairwiseMargins(final Set<String> winners,
+                                             final Set<String> losers,
+                                             final Map<String, Integer> voteTotals) {
+    final Set<Integer> margins = new HashSet<>();
+
+    if (losers.isEmpty()) {
+      margins.add(0);
+    } else {
+      for (final String w : winners) {
+        for (final String l : losers) {
+          margins.add(voteTotals.get(w) - voteTotals.get(l));
+        }
+      }
+    }
+
+    return margins;
+  }
+
   /**
    * Set voteTotals on CONTEST based on all counties that have that
    * Contest name in their uploaded CVRs
    **/
-  public static ContestResult countContest(final Map.Entry<String,
-                                           List<CountyContestResult>>
-                                           countyContestResults) {
-    final ContestResult contestResult =
-      ContestResultQueries.findOrCreate(countyContestResults.getKey());
+  public static ContestResult
+    countContest(final Map.Entry<String, List<CountyContestResult>> countyContestResults) {
+    final String contestName = countyContestResults.getKey();
+    final ContestResult contestResult = ContestResultQueries.findOrCreate(contestName);
 
     final Map<String,Integer> voteTotals =
       accumulateVoteTotals(countyContestResults.getValue().stream()
                            .map((cr) -> cr.voteTotals())
                            .collect(Collectors.toList()));
     contestResult.setVoteTotals(voteTotals);
-    // TODO how many winners are allowed?
-    contestResult.setWinners(winners(voteTotals));
+
+    int numWinners;
+    final Set<Integer> winnersAllowed = countyContestResults.getValue().stream()
+      .map(x -> x.winnersAllowed())
+      .collect(Collectors.toSet());
+
+    if (winnersAllowed.isEmpty()) {
+      LOGGER.error(String.format("[countContest: %s doesn't have any winners allowed."
```

Comment:
this kind of feels like a warning because a default of one seems reasonable.

Reply:
It's a (far-flung?) edge case, but we don't have any guarantees. Still 🤔 .